### DISCUSSION TITLE
How can I pass data between routes when using a redirect?

### DISCUSSION THREAD

**Discussion by greenshield** [vercel/next.js#53952] [App Router]
I have a question about passing data between routes using the app router.

I have a server action located at /src/lib/auth.js:

```
"use server";
import { redirect } from "next/navigation";

export async function sendReset(request) {

  //send reset email here

  redirect("/auth/signin");
}
```

And I have a password reset page located at /src/app/auth/reset/page.js which uses the sendReset server action as a formAction:

```
import { Input } from "@nextui-org/react";
import { sendReset } from "/../../lib/auth";

export default Reset(props) {
return <form action={sendReset}>

  <div>
    <Input label="Email" name="email" />
  </div>

  <div>
    <input type="submit" value="Reset Password">
  <div>

</form>
}

```

And I have a sign in page located at /src/app/auth/signin/page.js:

```
import { Input } from "@nextui-org/react";
import { signIn } from "/../../lib/auth";

export default async function SignIn(props) {
return <form action={signIn}>

  <div>
    <Input label="Email" name="email" />
  </div>

  <div>
    <Input label="Password" name="password" type="password />
  </div>

  <div>
    <input type="submit" value="Sign In">
  <div>

</form>
}
```

When I am on the password reset page and submit the form, the form action runs my server function sendReset() which sends the password reset link via email and then redirects to /auth/signin

This works exactly as I expect it to. So far so good.

However, I want to show a message on the SignIn page if the user arrived there after being redirected after a password reset request.

I know that I could just change my redirect from this:

```
redirect("/auth/signin");
```

to this:

```
redirect("/auth/signin?flash_message=sentreset");
```

And then detect the flash_message URL query param on /auth/signin and display a message based on the query param.

However, I don't want to show a query param in the URL.

I am able to use middleware to detect the flash_message query param from /auth/signin?flash_message=sentreset and then redirect to /auth/signin like this:

```
import { NextResponse } from "next/server";

export function middleware(request) {
    var nextUrl = request.nextUrl;

    var flash_message = request.nextUrl.searchParams.get("flash_message")

    nextUrl.search = "";
    return NextResponse.redirect(nextUrl);
}
```

But I don't know what to do with the flash_message variable in between so that it is accessible to me in the /auth/signin route.

I tried to set a header like this:

```
import { NextResponse } from "next/server";

export function middleware(request) {
  var nextUrl = request.nextUrl;

  var flash_message = request.nextUrl.searchParams.get("flash_message")

  if (flash_message) {
    nextUrl.search = "";
    return NextResponse.redirect(nextUrl, {
      headers: {
        "flash-message": "sentreset",
      },
    });
  }
}
```

This successfully detects the flash_message query param and sets the header, but the Flash-Message header only shows up in the Response Headers for the response that contains the 307 redirect (/auth/signin?flash_message=resetsent), but isn't passed to the url I'm redirecting to (/auth/signin)

I didn't think using headers was a great option anyway, but I wasn't sure what else to try.

Is there a flash storage option or something I can use to pass data between routes without showing the query string?

What's the right way to pass data between routes temporarily?

I want to be able to access the message parameter and also I want access to the value the user put into the email field on the reset page. How can I access these?

I considered trying to put the message in my redux store somehow, but I can't use redux in my server components, plus I was hoping to do this purely with the Next.js app router.

Thanks so much!

**Comment by Nadeera3784** [comment]
Hi @greenshield ,

since you want to avoid Redux how about using react context API?
for example :

```
// reset/page.js
import { Input } from '@nextui-org/react';
import { sendReset } from '/../../lib/auth';
import { useMessage } from '/path-to/MessageContext';
import { useRouter } from 'next/router';

export default function Reset(props) {
  const { setFlashMessage } = useMessage();
  const router = useRouter();

  const handleReset = async (e) => {
    e.preventDefault();
    // send reset email logic
    setFlashMessage('sentreset');
    router.push('/auth/signin');
  };

  return (
    <form onSubmit={handleReset}>
      {/* form contents */}
    </form>
  );
}


// signin/page.js
import { Input } from '@nextui-org/react';
import { useMessage } from '/path-to/MessageContext';

export default function SignIn(props) {
  const { flashMessage } = useMessage();

  return (
    <div>
      {flashMessage === 'sentreset' && (
        <p>Password reset email sent. Please check your inbox.</p>
      )}
      <form>
        {/* form contents */}
      </form>
    </div>
  );
}

```

**Reply by greenshield** [reply]
It's not that I don't want to use redux, it's that I can't use it in server components. Isn't it the same with useContext?

sendReset is a server action and redux / useContext require "use client"

**Comment by philwolstenholme** [comment]
How about writing to and reading from a cookie?
